### ISSUE TITLE
ENH: Discussion: Integrating Numpy with CPython interpreter specializations

### ISSUE THREAD

**Issue by mdboom** [numpy/numpy#27456]
### Proposed new feature or change:

There is an interesting recent paper about exposing the bytecode interpreter specialization features added in Python 3.11 (aka quickening) to arbitrary extension types.
 
https://drops.dagstuhl.de/entities/document/10.4230/LIPIcs.ECOOP.2024.6

For background, the bytecode specialization work for Python 3.11 essentially amounts to: if we notice that a generic instruction `BINARY_OP_ADD` is being called from a hot loop, replace it with an instruction that is specialized for the specific types that are present at the call site, e.g. `BINARY_OP_ADD_INT`.  This concept is the bulk of what gave Python 3.11 its ~30% speed improvement.

The innovation in the paper is to expose this facility to arbitrary extension types, so that a type can cache the results of dealing with dynamism at each call site and then skip that work on subsequent calls.  In the case of Numpy, this is all the "work planning" of handling casting, broadcasting, overrided ufuncs etc. that *usually* (but not always) doesn't change between calls at the same call site.  A key finding of the paper is that this overhead is enough, especially in the case of small to medium sized arrays, that the performance improvements can be very significant, on the order of more than 2x.

Prototype patches against CPython and Numpy used for the paper are available here:

Changes to Numpy: https://github.com/fberlakovich/cmq-numpy-ae
Changes to CPython: https://github.com/fberlakovich/cmq-ae

As a member of the Faster CPython team, I'd like to work on making this a reality, but the majority of the work actually needs to happen in Numpy, and I could use some help on that side and would also like to gauge the interest of the Numpy developers first.  (I've made modest contributions to Numpy in the past, but I definitely no longer have my bearings).  Looking at the patch to Numpy, you can see it involves generating code to handle and specialize many of the cases.  The authors used their own mako templates to do this, but they admit in the paper that it would perhaps be better to leverage the code generation that exists already.

What do you all think is the best way to proceed?  Would this require a NEP?

(For reference some [earlier discussion in the faster-cpython/ideas tracker](https://github.com/faster-cpython/ideas/issues/696)).

**Comment by mhvk** [comment]
This sounds very interesting! The paper is also nice and clear; less obvious is how to see what was actually changed in numpy (surely there is some handy feature that can show the changes made...). Ping @seberg, as he did quite a bit of ufunc caching, not all of which was in 1.26 (I think).

**Comment by mdboom** [comment]
The authors' changes to Numpy are all in this single commit on their fork: https://github.com/fberlakovich/cmq-numpy-ae/commit/301864ac4b85a49b8430ac5fb5382e663f7aabb9

(They admit in the paper that this is not the best way to do it, but it helped them quickly make a prototype).

**Comment by eendebakpt** [comment]
The paper was very nice to read and I would be interested in help out realizing this. One of the main questions I have is: how much of the gain can we realize in numpy internally? And how much can we then gain by integration with the cpython adaptive specialization?

I have experimented with numpy optimizations before (fast paths for dtype discovery, free lists for scalars) and there is room for improvement there.

**Comment by mdboom** [comment]
Thanks, I'm excited there is interest to collaborate on this.

I think the major gains from this approach would only happen when both CPython and Numpy are participating in adaptive specalization.  So that would be CPython 3.14 at the earliest.  There may of course be other ways to optimize how numpy does these things globally, but the "magic insight" in the paper is that by caching things at each particular call site you can do much better.

I think the easiest path forward would be for me to get @fberlakovich's fork on CPython working on top CPython main.  Once we have that, we can start on the work within Numpy itself -- basically it's to modify the generated ufuncs to handle saving/restoring cached work.  The existing fork hopefully serves as a guide, but (like the paper's authors suggest) maybe there is a simpler way than using those mako templates that are separate from the existing code generation in Numpy.

**Comment by fberlakovich** [comment]
Hi! 
Glad you liked the paper and that you are experimenting with it. I agree with @eendebakpt that NumPy probably also has internal room for optimization (the ufunc lookup machinery is quite complex), but I think most optimizations would need some sort of link to a specific occurence in the program to make assumptions about their context (e.g. the other side is always an int). Please let me know if you have any questions. I hope I will find time soon to repair the docker image to ease reproduction of the results.

**Comment by rgommers** [comment]
Quite interesting, thanks for sharing!

> A key finding of the paper is that this overhead is enough, especially in the case of small to medium sized arrays, that the performance improvements can be very significant, on the order of more than 2x.

I suspect that the single most important question to assess the benefits here is what "small to medium sized" means. The results in the paper don't say at all as far as I can tell. If there's code in working state, producing some graphs like this (from [this notebook](https://github.com/rgommers/explore-array-computing/blob/master/explore_xnd.ipynb)) for benchmarks as a function of input array size for benchmark problems that show larger speedups would be very helpful:

<img width="422" alt="image" src="https://github.com/user-attachments/assets/df6cf440-ddf1-4e64-a2f0-8458bfdc5c28">

If I understand correctly, most of the paper is about reducing constant overhead per expression. Whether the gain of that remains significant to O(100), O(1k) or O(10k) elements matters quite a lot. 

In addition, there is a memory reuse technique ("occurrence cache", section 4.2.3), the impact of which isn't analyzed separately, but that's something that potentially has benefits independent of input array size?

> Would this require a NEP?

Yes I think so, this seems significant enough that that is needed. In addition to making the case that the performance benefits are worth it compared to the costs of code complexity and potential other costs (e.g., binary size, build time, support for alternative interpreters), it would be nice to see something about topics like:
- interactions within NumPy, e.g., does it interact at all with the SIMD infrastructure)
- interactions outside of NumPy, e.g., with third-party dtypes, with Cython and with Numba
- how much brand-new Python C API will be used, and does that do anything to stability or downstream usage of the limited C API
    - I can also imagine that usage needs to be opt-in at build time for a while here, to avoid breaking users who for example build main branches of CPython and NumPy.

**Comment by seberg** [comment]
Just catching up with things.  I have to say, I am impressed by diving into and correctly describing the ufunc and indexing code paths! :)

I do have a few questions notes:
1. A large chunk of the ufunc improvement may be the result cache?  I am a bit worried that such a cache should not be persistent forever because it could waste memory indefinitely?  OTOH, I could imagine that the existing data cache should be improved (and maybe extended with an array free-list or some other optimization).
2. I do really like the indexing specialization example quite a lot and similarly, I wonder/assume that we can specialize things like `arr + 1` to `arr + np.array(1.0)` (caching the second one).
   Both are cases that add very significant overheads to operations (if the result is a small array or scalar).
3. Maybe the most intriguing to me, could be the mention of "Superinstructions".  There is recurring talk that NumPy could/should provide something similar to `numexpr`:  A way to combine multiple ufunc calls into a single "fused" ufunc.
   * If this could be automatically found (easily!) the story and use-case for such fused ufuncs would be even more compelling and in that case provide significant speed-ups to large arrays (rather than small to midsized ones as probably the case currently).

On the slightly cautious side, I will also say that for ufuncs, it may be that general useful optimizations in the ufunc code might reduce the usefulness of these optimizations a bit.

**Comment by fberlakovich** [comment]
> I suspect that the single most important question to assess the benefits here is what "small to medium sized" means. The results in the paper don't say at all as far as I can tell. If there's code in working state, producing some graphs like this (from [this notebook](https://github.com/rgommers/explore-array-computing/blob/master/explore_xnd.ipynb)) for benchmarks as a function of input array size for benchmark problems that show larger speedups would be very helpful:

I agree. For the paper we simply used the `paper` input sizes from the NPBench paper, but more detailed measurements would be interesting. Generally, a component analysis to figure out the impact of each optimization would be interesting, but we simply couldn't fit it in anymore before the paper deadline.

> * A large chunk of the ufunc improvement may be the result cache?  I am a bit worried that such a cache should not be persistent forever because it could waste memory indefinitely?  OTOH, I could imagine that the existing data cache should be improved (and maybe extended with an array free-list or some other optimization).

As with basically the entire implementation, there is almost certainly room for improvement, it is very researchy ;). There is, however, a simple cache invalidation heuristic that clears large cached arrays at the end of a function: https://github.com/fberlakovich/cmq-numpy-ae/blob/master/numpy/core/src/multiarray/number.c#L412. 

> 2\. I do really like the indexing specialization example quite a lot and similarly, I wonder/assume that we can specialize things like `arr + 1` to `arr + np.array(1.0)` (caching the second one).
>     Both are cases that add very significant overheads to operations (if the result is a small array or scalar).

I also like this one conceptually, but iirc the benefit compared to the other optimizations was low. I gathered this only from playing with the benchmarks, however. Only a sound component-wise analysis can tell for sure.

> * If this could be automatically found (easily!) the story and use-case for such fused ufuncs would be even more compelling and in that case provide significant speed-ups to large arrays (rather than small to midsized ones as probably the case currently).

Yes, that would be super cool. I also talked to @omelancon about possible benefits of combining our work (he had an excellent talk on static basic block versioning for Python, just before me).

**Comment by mhvk** [comment]
Agreed that superinstructions is a different beast, as ideally those would use some type of chaining of the ufunc inner loops (see https://github.com/mhvk/chain_ufunc for some experiments I did)
